File: app/src/main/java/ru/magflayer/spectrum/domain/injection/AppModule.java
package ru.magflayer.spectrum.domain.injection;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.google.gson.Gson;
import com.squareup.otto.Bus;
import com.squareup.otto.ThreadEnforcer;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import ru.magflayer.spectrum.data.database.AppDatabase;
import ru.magflayer.spectrum.data.system.LocalFileManager;
import ru.magflayer.spectrum.domain.repository.FileManagerRepository;
import ru.magflayer.spectrum.domain.repository.PhotoRepository;

@Module
class AppModule {

    private Context context;

    AppModule(Context context) {
        this.context = context;
    }

    @Provides
    Context provideContext() {
        return context;
    }

    @Singleton
    @Provides
    Bus provideBus() {
        return new Bus(ThreadEnforcer.ANY);
    }

    @Singleton
    @Provides
    Gson provideGson() {
        return new Gson();
    }

    @Singleton
    @Provides
    AppDatabase provideAppDatabase() {
        return Room.databaseBuilder(context, AppDatabase.class, AppDatabase.DATABASE_NAME)
                .build();
    }

    @Singleton
    @Provides
    PhotoRepository providePhotoRepository(final AppDatabase appDatabase) {
        return appDatabase;
    }

    @Singleton
    @Provides
    FileManagerRepository provideFileManagerRepository(final LocalFileManager localFileManager) {
        return localFileManager;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/domain/injection/AppModule.kt
package ru.magflayer.spectrum.domain.injection

import android.content.Context
import androidx.room.Room
import com.google.gson.Gson
import dagger.Module
import dagger.Provides
import ru.magflayer.spectrum.data.database.AppDatabase
import ru.magflayer.spectrum.data.database.ColorInfoRepositoryImpl
import ru.magflayer.spectrum.data.repository.PageAppearanceRepositoryImpl
import ru.magflayer.spectrum.data.repository.ToolbarAppearanceRepositoryImpl
import ru.magflayer.spectrum.data.system.LocalFileManager
import ru.magflayer.spectrum.domain.repository.*
import javax.inject.Singleton

@Module
internal class AppModule(private val context: Context) {

    @Provides
    fun provideContext(): Context {
        return context
    }

    @Singleton
    @Provides
    fun provideGson(): Gson {
        return Gson()
    }

    @Singleton
    @Provides
    fun provideAppDatabase(): AppDatabase {
        return Room.databaseBuilder(context, AppDatabase::class.java, AppDatabase.DATABASE_NAME)
            .fallbackToDestructiveMigration()
            .build()
    }

    @Singleton
    @Provides
    fun providePhotoRepository(appDatabase: AppDatabase): PhotoRepository {
        return appDatabase
    }

    @Singleton
    @Provides
    fun provideFileManagerRepository(localFileManager: LocalFileManager): FileManagerRepository {
        return localFileManager
    }

    @Singleton
    @Provides
    fun provideColorInfoRepository(appDatabase: AppDatabase): ColorInfoRepository {
        return ColorInfoRepositoryImpl(appDatabase)
    }

    @Singleton
    @Provides
    fun provideToolbarAppearanceRepository(): ToolbarAppearanceRepository {
        return ToolbarAppearanceRepositoryImpl()
    }

    @Singleton
    @Provides
    fun providePageAppearanceRepository(): PageAppearanceRepository {
        return PageAppearanceRepositoryImpl()
    }
}
